DPPP/MSBDAReader.h

```diff
@@ -0,0 +1,203 @@
+///       They are read from the FLAG column. If a FLAG_ROW is set, all
+///       flags for that baseline will be set. Also the flag of data
+///       containing NaN or infinite numbers will be set.
+///       All flags of an inserted time slot are set.
+///   </td>
+///  </tr>
+///  <tr>
+///   <td>WEIGHT</td>
+///   <td>The data weights as [ncorr,nchan,nbaseline]. Column
+///       WEIGHT_SPECTRUM is used if present and containing valid data,
+///       otherwise column WEIGHT is used. The weights of an inserted
+///       time slot are set to 0.
+///       If autoweight is on, the autocorrelations are used to
+///       calculate proper weights.
+///   </td>
+///  </tr>
+///  <tr>
+///   <td>UVW</td>
+///   <td>The UVW coordinates in meters as [3,nbaseline].
+///       They are calculated for a missing time slot.
+///   </td>
+///  </tr>
+///  <tr>
+///   <td>FULLRESFLAG</td>
+///   <td>For each baseline the LOFAR_FULL_RES_FLAG column is stored as
+///       a uChar array with shape [orignchan/8, ntimeavg]. The bits
+///       represent the flags. They are converted to a Bool array with shape
+///       [orignchan, ntimeavg, nbaseline].
+///       If column LOFAR_FULL_RES_FLAG is not present, the flags are used
+///       and it is assumed that no averaging was done yet (thus ntimeavg=1
+///       and orignchan=nchan).
+///   </td>
+///  </tr>
+/// </table>
+
+class MSBDAReader : public DPInput {
+ public:
+  /// Default constructor.
+  MSBDAReader();
+
+  /// Construct the object for the given MS.
+  /// Parameters are obtained from the parset using the given prefix.
+  MSBDAReader(const std::string& msName, const ParameterSet&,
+              const std::string& prefix);
+
+  virtual ~MSBDAReader();
+
+  /// Process the next data chunk.
+  /// It returns false when at the end.
+  bool process(const DPBuffer&) override;
+
+  /// Finish the processing of this step and subsequent steps.
+  void finish() override;
+
+  /// Return which datatype this step outputs.
+  MSType outputs() const override;
+
+  /// Update the general info.
+  void updateInfo(const DPInfo&) override;
+
+  /// Add some data to the MeasurementSet written/updated.
+  /// Do nothing.
+  void addToMS(const string&) override{};
+
+  /// Show the step parameters.
+  void show(std::ostream&) const override;
+
+  /// If needed, show the flag counts.
+  void showCounts(std::ostream&) const override;
+
+  /// Show the timings.
+  void showTimings(std::ostream&, double duration) const override;
+
+  /// Get the name of the MS.
+  std::string msName() const override;
+
+  /// Tell if the visibility data are to be read. If set to true once,
+  /// this will stay true.
+  void setReadVisData(bool readVisData) override;
+
+  /// Get the main MS table.
+  casacore::Table& table() override { return ms_; }
+
+  /// Get the selected spectral window.
+  unsigned int spectralWindow() const override { return spw_; }
+
+ protected:
```

Comment:
Using protected members is almost as evil as using public members: Any derived class can modify them, so there's no guarantee anymore about the validity of these members. Please make them private, or clearly document why we absolutely needs protected members.